File: algorithms/solutions/java/_0002_AddTwoNumbers.java
package hot100;

public class _0002_AddTwoNumbers {
    // Definition for singly-linked list.
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
            next = null;
        }
    }

// @lc code=start
    class Solution {
        public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
            ListNode dummy = new ListNode(-1);
            ListNode cur = dummy;
            int value = 0, carry = 0;
            while (l1 != null || l2 != null) {
                int v1 = l1 != null ? l1.val : 0;
                int v2 = l2 != null ? l2.val : 0;
                value = (v1 + v2 + carry) % 10;
                carry = (v1 + v2 + carry) / 10;
                ListNode node = new ListNode(value);
                cur.next = node;
                cur = node;
                if (l1 != null) {
                    l1 = l1.next;
                }
                if (l2 != null) {
                    l2 = l2.next;
                }
            }
            if (carry != 0) {
                cur.next = new ListNode(carry);
            }
            return dummy.next;
        }
    }
// @lc code=end

}

File: algorithms/solutions/python3/0002_add_two_numbers.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] Add Two Numbers
#
# https://leetcode.cn/problems/add-two-numbers/description/
#

# @lc code=start
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(-1)
        cur = dummy
        value = carry = 0
        while l1 or l2:
            v1 = 0 if not l1 else l1.val
            v2 = 0 if not l2 else l2.val
            value = (v1 + v2 + carry) % 10
            carry = (v1 + v2 + carry) // 10
            node = ListNode(value)
            cur.next = node
            cur = node
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        if carry != 0:
            cur.next = ListNode(carry)
        return dummy.next
# @lc code=end

